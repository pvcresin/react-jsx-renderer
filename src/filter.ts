import { JSXFilter } from "./evaluator";

export const DenyAttributeFilter: (matchers: (string | RegExp)[]) => JSXFilter = (matchers) => {
  const regexps = matchers.map((matcher) => matcher instanceof RegExp ? matcher : new RegExp(matcher));
  return (element) => {
    const { props, ...spread } = element;

    const filteredProps = { ...props };
    Object.keys(filteredProps).forEach((key) => {
      if(regexps.reduce<Boolean>((match, regexp) => match || regexp.exec(key), false)) {
        delete filteredProps[key];
      }
    })

    return { ...spread, props: filteredProps }
  }
}
