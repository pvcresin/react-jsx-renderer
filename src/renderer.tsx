import { Component, createElement, ReactNode } from "react";
import { JSXElement, JSXEvaluator, JSXEvaluatorOption } from "./evaluator";

export interface JSXRendererProps extends Partial<JSXEvaluatorOption> {
  code: string;
  fallback?: (error: Error) => JSX.Element | undefined;
}

interface State {}

export class JSXRenderer extends Component<JSXRendererProps, State> {
  static displayName = 'JSX';

  constructor(props: Readonly<JSXRendererProps>) {
    super(props);

    this.state = {};
  }

  public render() {
    try {
      const chidlren = this.eval(this.props.code);
      return <>{chidlren}</>;
    } catch(err) {
      const error = err instanceof Error ? err : new Error(err.toString());
      return this.props.fallback ? this.props.fallback(error) : <>{error.message}</>;
    }
  }

  private eval(code: string): ReactNode[] {
    const evaluator = new JSXEvaluator({
      binding: this.props.binding || {},
      components: this.props.components || {},
      filters: this.props.filters || [],
    });
    return evaluator.eval(code).map((element) => this.toReactNode(element))
  }

  private toReactNode(element: JSXElement | string | number | boolean): ReactNode {
    if (typeof element === 'object') {
      return createElement(element.component, element.props, element.children.map((child) => this.toReactNode(child)))
    }
    return element;
  }
}
