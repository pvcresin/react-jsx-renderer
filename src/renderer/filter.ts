import { JSXElement, JSXFragment, JSXText } from '../types';

export type JSXElementFilter = (node: JSXElement) => JSXElement | undefined;
export type JSXFragmentFilter = (node: JSXFragment) => JSXFragment | undefined;
export type JSXTextFilter = (node: JSXText) => JSXText | undefined;
