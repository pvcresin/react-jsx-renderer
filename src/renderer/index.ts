export * from './options';
export * from './filter';
export * from './renderer';
