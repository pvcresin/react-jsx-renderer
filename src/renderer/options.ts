import { JSXElementFilter, JSXFragmentFilter, JSXTextFilter } from "./filter";

export interface RenderingOptions {
  disableUnknownHTMLElement?: boolean;

  elementFilters?: JSXElementFilter[];
  fragmentFilters?: JSXFragmentFilter[];
  textFilters?: JSXTextFilter[];
}
