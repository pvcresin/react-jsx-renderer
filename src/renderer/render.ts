import { createElement, Fragment, ReactNode } from 'react';
import { JSXNode, JSXElement, JSXFragment, JSXText } from '../types';
import { RenderingOptions } from './options';

const unknownElementCache = new Map<string, boolean>();

export const render = (node: JSXNode, options: RenderingOptions): ReactNode => {
  if (node === null) return node;
  if (node === undefined) return node;

  switch (typeof node) {
    case 'boolean':
      return node;
    case 'string':
    case 'number':
      return renderJSXText(node, options);
    default:
      return renderJSXNode(node, options);
  }
};

const renderJSXText = (text: JSXText, options: RenderingOptions): ReactNode => {
  return applyFilter(options.textFilters || [], text);
};

const renderJSXNode = (node: JSXElement | JSXFragment, options: RenderingOptions): ReactNode => {
  switch (node.type) {
    case 'element':
      return renderJSXElement(node, options);
    case 'fragment':
      return renderJSXFragment(node, options);
  }
};

const renderJSXElement = (element: JSXElement, options: RenderingOptions): ReactNode => {
  const filtered = applyFilter(options.elementFilters || [], element);
  if (!filtered) return undefined;

  if (options.disableUnknownHTMLElement && typeof filtered.component === 'string') {
    const { component } = filtered;
    if (!unknownElementCache.has(component)) {
      unknownElementCache.set(component, document.createElement(component) instanceof HTMLUnknownElement);
    }
    if (unknownElementCache.get(component)) return undefined;
  }

  return createElement(filtered.component, filtered.props, ...filtered.children.map((child) => render(child, options)));
};

const renderJSXFragment = (fragment: JSXFragment, options: RenderingOptions): ReactNode => {
  const filtered = applyFilter(options.fragmentFilters || [], fragment);

  if (filtered) {
    return createElement(Fragment, filtered.props, ...filtered.children.map((child) => render(child, options)));
  } else {
    return undefined;
  }
};

const applyFilter = <T extends JSXNode>(filters: ((target: T) => T | undefined)[], node: T): T | undefined => {
  return filters.reduce<T | undefined>((prev, filter) => (prev ? filter(prev) : undefined), node);
};
