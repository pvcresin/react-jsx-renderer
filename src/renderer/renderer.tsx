import { memo, VFC } from "react";
import { evaluateJSX, EvaluateOptions } from "../evaluate";
import { RenderingOptions } from "./options";
import { render } from "./render";

export type JSXFallbackComponent = VFC<{ error: Error }>;

export interface JSXRendererProps extends EvaluateOptions, RenderingOptions {
  code: string;
  fallbackComponent?: JSXFallbackComponent;
}

export const JSXRenderer: VFC<JSXRendererProps> = memo((props) => {
  const { code, ...options } = props;

  try {
    const children = evaluateJSX(code, options);
    return <>{children.map((child) => render(child, options))}</>;
  } catch(err) {
    const error = err instanceof Error ? err : new Error(err);
    const Fallback = options.fallbackComponent ? options.fallbackComponent : ({ error }) => <>{error.message}</>;
    return <Fallback error={error} />
  }
})
