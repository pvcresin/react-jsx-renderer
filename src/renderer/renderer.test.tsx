import { ReactElement } from 'react';
import renderer from 'react-test-renderer';
import { JSXRenderer } from '.';
import { JSXElementFilter, JSXFragmentFilter, JSXTextFilter } from './filter';
import { JSXFallbackComponent } from './renderer';

describe('JSX', () => {
  it('should render JSX', () => {
    const tree = renderer
      .create(
        <JSXRenderer
          code={`
      <h1>Hello, JSX World</h1>
      <p>This is paragraph.</p>
      <ul className="Hello">
        <li>first</li>
        <li>second</li>
      </ul>
      <>Fragment</>
    `}
          keyPrefix="html"
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should catch error of syntax', () => {
    const tree = renderer.create(<JSXRenderer disableKeyGeneration code={`<p ...`} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should catch error with number object', () => {
    const tree = renderer
      .create(
        <JSXRenderer
          disableKeyGeneration
          code={`<p>{raise()}</p><p />`}
          binding={{
            raise: () => {
              throw 1;
            },
          }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should catch error with object', () => {
    const tree = renderer
      .create(
        <JSXRenderer
          disableKeyGeneration
          code={`<p>{raise()}</p>`}
          binding={{
            raise: () => {
              throw { foo: 1 };
            },
          }}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  const elementFilters: JSXElementFilter[] = [(e) => (e.component === 'link' ? undefined : e), (e) => (e.component === 'script' ? undefined : e)];
  const fragmentFilters: JSXFragmentFilter[] = [
    (fragment) => {
      return fragment.children[0] === 'delete' ? undefined : fragment;
    },
  ];
  const textFilters: JSXTextFilter[] = [(text) => (text === 'ban' ? undefined : text)];
  const fallback: JSXFallbackComponent = ({ error }) => <div>{error.message} is raised</div>;
  const test = (code: string, expected: ReactElement) => {
    it(`should render ${code}`, () => {
      const actual = renderer
        .create(
          <JSXRenderer
            code={code}
            binding={{
              exception() {
                throw new Error('Hello');
              },
            }}
            elementFilters={elementFilters}
            fragmentFilters={fragmentFilters}
            textFilters={textFilters}
            fallbackComponent={fallback}
            disableSearchCompontsByBinding
            disableUnknownHTMLElement
          />,
        )
        .toJSON();
      expect(actual).toStrictEqual(renderer.create(expected).toJSON());
    });
  };

  test('', <></>);
  test('<p>Hoge</p>', <p>Hoge</p>);
  test('<p>{true}</p>', <p>{true}</p>);
  test('<p>{null}</p>', <p>{null}</p>);
  test('<p>{undefined}</p>', <p>{undefined}</p>);
  test(
    '<p><>frag</></p>',
    <p>
      <>frag</>
    </p>,
  );
  test('<p><>delete</></p>', <p></p>);
  test('<p><script src="./index.js" /></p>', <p></p>);
  test('<p><link href="./index" /></p>', <p></p>);
  test('<p><unknown>this is removed</unknown>element</p>', <p>element</p>);
  test(
    '<p><b>ban</b>word</p>',
    <p>
      <b />
      word
    </p>,
  );
  test('<p>{exception()}</p>', <div>{'Hello'} is raised</div>);
});
