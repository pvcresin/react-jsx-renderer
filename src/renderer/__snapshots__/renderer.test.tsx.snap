// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSX should catch error of syntax 1`] = `"[1:11]: Unexpected token"`;

exports[`JSX should catch error with number object 1`] = `"1"`;

exports[`JSX should catch error with object 1`] = `"[object Object]"`;

exports[`JSX should render JSX 1`] = `
Array [
  "
      ",
  <h1>
    Hello, JSX World
  </h1>,
  "
      ",
  <p>
    This is paragraph.
  </p>,
  "
      ",
  <ul
    className="Hello"
  >
    
        
    <li>
      first
    </li>
    
        
    <li>
      second
    </li>
    
      
  </ul>,
  "
      ",
  "Fragment",
]
`;
